README.md

```diff
@@ -1,161 +1,720 @@
+
+An "entity" MUST represent some subject in the database that can have an attribute. Since names are not unique (and are in fact an attribute), each entity needs a unique identity. Using a random number of at least 128-bits when generating a fresh entity is is RECOMMENDED.
+
+``` typescript
+type EntityID = Bytes
+```
+
+Especially since PomoDB allows for creation IDs without coordination, a real-world entity MAY be represented by multiple entity IDs. This case SHOULD be handed at the query layer by selecting for all relevant entity IDs.
+
+### 4.1.4 Attribute
+
+Attributes MUST be an integer, double-precision float, UTF8 string, or binary:
+
+``` typescript
+type Attribute
+  = Integer // e.g. Normal indices
+  | Double  // e.g. Fractional indices
+  | Utf8
+  | Bytes
+```
+
+### 4.1.5 Value
+
+The EDB MUST support the following primitive value types:
+
+``` typescript
+type Value
+  = Boolean
+  | Integer
+  | Double
+  | Utf8
+  | Cid
+  | Bytes
+```
+
+### 4.1.6 Cause
+
+Causal links MUST be expressed as an unordered set of CIDs, where those CIDs are other PomoDB fact CIDs. CIDs MUST be deduplicated at read time (though deduplicating and sorting at write time is RECOMMENDED).
+
+``` typescript
+type Cause = Set<Cid>
+```
+
+## 4.2 Graph Topologies
+
+All relations in PomoDB MAY be expressed in several ways, but common terminology around graphs is RECOMMENDED as it is very clear and pictorial.
+
+### 4.2.1 Grouping: Set & Stars
+
+Grouping by entity ID, attribute, or value all produce sets. In graph terms, this is expressed as a "star" or "hub and spoke" topology.
+
+| Role  | Description                                              |
+|-------|----------------------------------------------------------|
+| Hub   | The center of the relation / the item being "grouped by" |
+| Spoke | The thing being related                                  |
+
+``` mermaid
+flowchart RL
+    sun(("☀️"))
+
+    earth(("🌏"))
+    saturn(("🪐"))
+    alien(("👽"))
+    uap(("🛸"))
+    meteor(("☄️"))
+
+    earth --- sun
+    sun --- saturn
+    alien --- sun
+    sun --- uap
+    meteor --- sun
+```
+
+Multiple related hubs are possible.
+
+``` mermaid
+flowchart RL
+    sun(("☀️"))
+
+    earth(("🌏"))
+    saturn(("🪐"))
+    alien(("👽"))
+    uap(("🛸"))
+    meteor(("☄️"))
+
+    earth ------ sun
+    sun --- saturn
+    alien --- sun
+    sun --- uap
+    meteor --- sun
+
+    moon(("🌖")) --- earth
+    satelite(("🛰️")) --- earth
+    earth --- astro(("👩‍🚀"))
+```
+
+### 4.2.2 Ordering & Causality
+
+An important order (or "sort") in PomoDB is causal order. This builds up a graph of pointers in the [`Cause`] field.
+
+Below is a example causal graph from three writers. The grouping in this diagram is not important, but is presented this way here for explanatory purposes. In reality, if two writers commit the same fact to the database with the same causal history, they would be be deduplicated since they would have the same CID.
+
+``` mermaid
```

Comment:
What do the colors in this diagram mean?

Reply:
They're referenced in the text below

Reply:
For example:

> The "earliest" facts in the graph above are `bafy...almond` and `bafy...bacon` (**highlighted in blue**),

(Also: I was tempted to try to colourize the word "house" in blue somewhere in the prose, but alas GitHub Flavoured Markdown doesn't support styling text 😭)